### ISSUE TITLE
passing part names to the install command is an attractive nuisance

### ISSUE THREAD

**Issue by jimfulton** [buildout/buildout#358]
Because the generated parts have unpredictable configuration, because the configuration of parts they reference may not be up to date.

Perhaps this feature could be rehabilitated by initializing, but not installing referenced parts.

Or perhaps we should just drop this dubious feature in buildout 3.  For now, I'm removing mention of the feature from the help and not including it in docs.

**Comment by leorochael** [comment]
Can you give a more concrete example of the problems this cause?

**Comment by jimfulton** [comment]
Not off the top of my head, but I've been burned by this in the past and know to ignore it. I think the problem is that references parts either aren't initialized correctly, or perhaps because the dependent part may depend on installation of the referenced parts.

It might be enough to modify the behavior to install only the named parts and the parts they transitively reference.  But generally, I think this feature invites inconsistency of various kinds.

**Comment by leorochael** [comment]
I see.

I thought the parts referenced by the mentioned part were indeed initialized implicitly by being referenced already, but if they're not installed, this could indeed result in a buggy installation if the installed parts depend on the mentioned parts being actually installed.

I agree that all the transitively referenced parts should be installed along with the ones mentioned, though I guess that should result in added logging to make it clear why a non-mentioned part was installed.

**Comment by jimfulton** [comment]
On Sun, Feb 26, 2017 at 2:42 PM, Leonardo Rochael Almeida <
notifications@github.com> wrote:

> I see.
>
> I thought the parts referenced by the mentioned part were indeed
> initialized implicitly by being referenced already, but if they're not
> installed, this could indeed result in a buggy installation if the
> installed parts depend on the mentioned parts being actually installed.
>
> I agree that all the transitively referenced parts should be installed
> along with the ones mentioned, though I guess that should result in added
> logging to make it clear why a non-mentioned part was installed.
>
But really, is it worth the bother?  Do you think people use this?

**Comment by leorochael** [comment]
> But really, is it worth the bother?

Not sure. I agree that if it's to be done, it should be done "properly" (installing transitive dependencies). I'd be -0 on dropping the feature.

> Do you think people use this?

I use it on occasion. One specific example (in a closed repo, unfortunately), is that in a CI pipeline (like travis) that allows me to run certain build steps in parallel, I run a `lint` step in parallel that starts with `bin/buildout install lint`, which only installs `pylint` and `flake8` eggs with their scripts, since none of the other (time consuming) steps are needed to check if the source code passes lint tests.

The `lint` part in the example above doesn't depend on any other part, so it's a case where  `buildout install part` just works. It might not be a representative example enough to argue for keeping the feature.

On the other hand, in the history we do mention the replacement of Make with a Python based alternative as an initial motivation for Buildout, and when developers (read: I) think of `Make replacement` they (i.e.: I) usually expect being able to run just some parts of the build (and their transitive dependencies).

**Comment by mauritsvanrees** [comment]
Some packages specify command line options in a `.travis.yml` file. For example [experimental.publishtraverse](https://github.com/plone/experimental.publishtraverse/blob/1.1/.travis.yml#L11):
```
bin/buildout -Nc travis.cfg buildout:eggs-directory=$HOME/buildout-cache/eggs \
    buildout:download-cache=$HOME/buildout-cache/downloads
```

Similar for [five.customerize(https://github.com/zopefoundation/five.customerize/blob/master/.travis.yml#L14):

```
bin/buildout -N buildout:download-cache=downloads
```

Those are the two I could find on my computer.

That could be done with an extra buildout config file for Travis. But that can also be yet another config file in an already overcrowded directory, so being able to override settings on the command line can be useful.

Well, [collective.recipe.plonesite](https://pypi.python.org/pypi/collective.recipe.plonesite#upgrading) mentions the possibility of basically switching that part off temporarily by running `bin/buildout -Nv plonesite:enabled=false`. It is a recipe that during the buildout run starts up a Zope instance, creates a Plone Site if it is not there yet, possibly runs some upgrades, and stops Zope again. That can take irritatingly long if you know that the site is already there and there is nothing to do, so switching it off from the command line can be very handy.
Also, you could pass `plonesite:site-id=newsite` on the command line to create a different site.

With the last example, what I *expect* to happen, is:
- buildout reads the config normally, without initializing or installing anything yet.
- buildout looks for the `plonesite` section and adds or edits the `site-id` option.
- buildout initializes and installs any parts that it would normally do. Just because the `plonesite` part is mentioned on the command line, does not mean that the `plonesite` part will be initialized or installed.

I don't know if that is what actually happens, but that is what I expect.

I guess where I can most easily see this go wrong, is when you use the command line to change an option for an extension: would buildout change the config before or after loading the extensions? I would say: before. Also, if you use `buildout:extends=some.cfg` it might also become tricky.

Long story short: there are uses. Not many though.  So if it is causing too many headaches to maintain, it can be something to get rid of.

**Comment by mgedmin** [comment]
Would `bin/buildout parts="X Y Z"` be a sufficient replacement for `bin/buildout install X Y Z`?  I think it would handle dependencies correctly.

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 7:52 AM, Maurits van Rees <notifications@github.com>
wrote:

> Some packages specify command line options in a .travis.yml file. For
> example experimental.publishtraverse
> <https://github.com/plone/experimental.publishtraverse/blob/1.1/.travis.yml#L11>
> :
>
> bin/buildout -Nc travis.cfg buildout:eggs-directory=$HOME/buildout-cache/eggs \
>     buildout:download-cache=$HOME/buildout-cache/downloads
>
> Similar for [five.customerize(https://github.com/zopefoundation/
> five.customerize/blob/master/.travis.yml#L14):
>
> bin/buildout -N buildout:download-cache=downloads
>
> Those are the two I could find on my computer.
>
I think you're misunderstanding the issue.

Command-line options *aren't* an attractive nuisance.

Command-line options to the *install command are*.

...

> Long story short: there are uses. Not many though. So if it is causing too
> many headaches to maintain, it can be something to get rid of.
>
I'm a huge fan of command-line options used to override the configuration
database.

That's not what this issue is about.

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 8:02 AM, Marius Gedminas <notifications@github.com>
wrote:

> Would bin/buildout parts="X Y Z" be a sufficient replacement for bin/buildout
> install X Y Z? I think it would handle dependencies correctly.
>
IDK, it depends on what problem you're trying to solve.

  install X Y Z

Installs/updates X, Y, and Z, leaving all other parts unchanged.  This can
lead to mysterious bugs because dependencies of X, Y, and Z are unaffected,

OTOH:

  bin/buildout parts="X Y Z"

Installs parts X, Y, Z, and their dependencies and uninstalls anything else
that was previously installed. This is very different.

I think we should step back and ask:

   1. What problem is this feature solving and
   2. Is the problem important enough for the feature to exist, recognizing
   that features are a liability because they:
   1. impose an implementation and documentation maintenance burden, and
      2. impose a burden on users learning buildout, if only giving them
      something to learn enough about to know decide whether they need to
      understand it.

The feature as it exists today is certainly an attractive nuisance because
it invites unpredictable outcomes.  It could be fixed, for some definition
of fixed, but I don't know of any benefit that would justify the bother.

**Comment by jimfulton** [comment]
TIL that github removes or mangles all attempts to format email responses nicely. :)

**Comment by mauritsvanrees** [comment]
Ah, indeed I misunderstood.

I can't remember ever having `bin/buildout install some part` go wrong, except for obvious reasons.

Yesterday I ran `bin/buildout install instance` a couple of times, because I removed or added a couple of eggs in the `instance` part. So I knew that the resulting `bin/instance` script was the only thing that could change. Or if other parts also changed, I didn't care: I wasn't going to use them at that point. Installing all parts would take longer than I wanted.

If the `instance` part references another part, I would expect that part to be initialized and installed (or updated). From what I read above, it seems that a referenced part is not installed, and maybe not even initialized. That would surprise me.

So: the main attraction of this feature for me would be: save time.

**Comment by mgedmin** [comment]
> Installs parts X, Y, Z, and their dependencies and uninstalls anything else
that was previously installed. 

Ouch, ouch, ouch, I didn't realize it would uninstall all the other parts!

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 8:53 AM, Marius Gedminas <notifications@github.com>
wrote:

> Installs parts X, Y, Z, and their dependencies and uninstalls anything else
> that was previously installed.
>
> Ouch, ouch, ouch, I didn't realize it would uninstall all the other parts!
>
It's just as if you edited the parts option in your buildout.

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 8:36 AM, Maurits van Rees <notifications@github.com>
wrote:

> Ah, indeed I misunderstood.
>
> I can't remember ever having bin/buildout install some part go wrong,
> except for obvious reasons.
>
> Yesterday I ran bin/buildout install instance a couple of times, because
> I removed or added a couple of eggs in the instance part. So I knew that
> the resulting bin/instance script was the only thing that could change.
> Or if other parts also changed, I didn't care: I wasn't going to use them
> at that point. Installing all parts would take longer than I wanted.
>
But why would they?  Are these using recipes with expensive update methods?

In my experience, running -N is the main time saver.  If parts don't have
to be reinstalled, they rarely take significant time. If they do, then
that's the problem, IMO.

> If the instance part references another part, I would expect that part to
> be initialized and installed (or updated). From what I read above, it seems
> that a referenced part is not installed, and maybe not even initialized.
> That would surprise me.
>
> So: the main attraction of this feature for me would be: save time.
>
I worry that this is inviting false optimizations.  At best, it seems to me
to invite people to focus on things that are unimportant.

**Comment by mauritsvanrees** [comment]
In this project, as in most of my projects, I have all packages pinned, so running in non-newest mode should not matter.

This project isn't big, just six parts: zeoclient (Plone site), zeoserver, backup (generates a couple of ZODB backup scripts), instance (almost same as zeoclient), ipzope (script with the eggs of the instance), omelette. Especially when everything has already been built once, it shouldn't take long.

The main `instance` part installs Plone 4.3, which means about 270 eggs. That is quite a lot, but they have already been downloaded and installed previously. This is with zc.buildout 2.8.0 and `abi-tag-eggs = true`, with Python 2.7 on Mac.

Let's see, I ran `bin/buildout` once, and then I ran it a couple of times with various options:

- `bin/buildout`: 35 seconds
- `bin/buildout -N`: 35 seconds
- `bin/buildout install instance`: 16 seconds

So installing everything takes twice as long.
Those are times reported as 'real' by `time`.  The 'user' times were one second shorter.
It takes about 8 seconds before buildout starts installing a part: it needs that time at least to read the config (nothing external) and develop a package, an probably initialize one or more parts.

It can be that some recipes could be made faster for updating, especially `plone.recipe.zope2instance` may be doing too much.

BTW, in this particular case, it would have been fine for me to use `parts = instance`.

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 9:40 AM, Maurits van Rees <notifications@github.com>
wrote:

> In this project, as in most of my projects, I have all packages pinned, so
> running in non-newest mode should not matter.
>
> This project isn't big, just six parts: zeoclient (Plone site), zeoserver,
> backup (generates a couple of ZODB backup scripts), instance (almost same
> as zeoclient), ipzope (script with the eggs of the instance), omelette.
> Especially when everything has already been built once, it shouldn't take
> long.
>
> The main instance part installs Plone 4.3, which means about 270 eggs.
> That is quite a lot, but they have already been downloaded and installed
> previously. This is with zc.buildout 2.8.0 and abi-tag-eggs = true, with
> Python 2.7 on Mac.
>
> Let's see, I ran bin/buildout once, and then I ran it a couple of times
> with various options:
>
>    - bin/buildout: 35 seconds
>    - bin/buildout -N: 35 seconds
>    - bin/buildout install instance: 16 seconds
>
> So installing everything takes twice as long.
> Those are times reported as 'real' by time. The 'user' times were one
> second shorter.
>
Recently I realized that passing

  -v log-format="%(asctime)s %(message)s"

Can be very helpful for getting a better handle on where time is spent.


> It takes about 8 seconds before buildout starts installing a part: it
> needs that time at least to read the config (nothing external) and develop
> a package, an probably initialize one or more parts.
>
I'm wildly skeptical that that is taking 8 seconds.

> It can be that some recipes could be made faster for updating, especially
> plone.recipe.zope2instance may be doing too much.
>
> BTW, in this particular case, it would have been fine for me to use parts
> = instance.
>
So you would have been cool with losing your ZEO server?

**Comment by jimfulton** [comment]
If someone wants to draft documentation for this feature and, based on that, at least a few of us think it's a good idea and then if someone is willing to implement it correctly, then I'll defer. :)

**Comment by mauritsvanrees** [comment]
> I'm wildly skeptical that that is taking 8 seconds.

It seems more the fault of the recipe after all. Or I misinterpreted some numbers.
A buildout with one small recipe takes about 2.5 seconds, also when it is part of a larger buildout and is being run with `bin/buildout install small-part`.

The logging is useful indeed. But the first two seconds it logs nothing at all, even with `-vvvvvvvvvvvv`. I don't know what buildout is doing then. Anyway, that is just two seconds, not eight, so I won't complain.

One more data point: when I use a recipe that needs `setuptools`, I get another 2 second pause while buildout is apparently busy with extra `find-links`:

```
2017-03-03 20:27:28,276 Getting required 'setuptools==34.3.0'
2017-03-03 20:27:28,277 We have the distribution that satisfies 'setuptools==34.3.0'.
2017-03-03 20:27:28,277 Adding find link 'https://files.pythonhosted.org/packages/source/c/certifi/certifi-2016.9.26.tar.gz#md5=baa81e951a29958563689d868ef1064d' from setuptools 34.3.0
2017-03-03 20:27:30,259 Adding find link 'https://files.pythonhosted.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2' from setuptools 34.3.0
```

Anyway, I am totally distracting this issue from a discussion about a feature into a debugging session about the speed of buildout or recipes. Let me shut up now.

**Comment by leorochael** [comment]
> Anyway, I am totally distracting this issue from a discussion about a feature into a debugging session about the speed of buildout or recipes. Let me shut up now.

Well, speed seems to be the main justification for keeping or not this feature, so I think a small digression into profiling is useful.

Anyway, we should do a [flame](https://github.com/evanhempel/python-flamegraph) [graph](https://eng.uber.com/pyflame/) of a Buildout run to see where the time is going.

**Comment by leorochael** [comment]
Also: https://github.com/nvdv/vprof

**Comment by jimfulton** [comment]
On Fri, Mar 3, 2017 at 2:33 PM, Maurits van Rees <notifications@github.com>
wrote:

> I'm wildly skeptical that that is taking 8 seconds.
>
> It seems more the fault of the recipe after all. Or I misinterpreted some
> numbers.
> A buildout with one small recipe takes about 2.5 seconds, also when it is
> part of a larger buildout and is being run with bin/buildout install
> small-part.
>
I wonder if it's checking for new versions of buildout and setuptools. Did
you use -N?

FWIW, running the ZODB buildout (not that small, but not Plone :)) takes a
little less than a second for me with -N. <shrug>


> The logging is useful indeed. But the first two seconds it logs nothing at
> all, even with -vvvvvvvvvvvv.
>
Note that more than 2 vs has no effect:

http://docs.buildout.org/en/latest/reference.html#buildout-command-line-options


> I don't know what buildout is doing then. Anyway, that is just two
> seconds, not eight, so I won't complain.
>
> One more data point: when I use a recipe that needs setuptools, I get
> another 2 second pause while buildout is apparently busy with extra
> find-links:
>
> 2017-03-03 20:27:28,276 Getting required 'setuptools==34.3.0'
> 2017-03-03 20:27:28,277 We have the distribution that satisfies 'setuptools==34.3.0'.
> 2017-03-03 20:27:28,277 Adding find link 'https://files.pythonhosted.org/packages/source/c/certifi/certifi-2016.9.26.tar.gz#md5=baa81e951a29958563689d868ef1064d' from setuptools 34.3.0
> 2017-03-03 20:27:30,259 Adding find link 'https://files.pythonhosted.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2' from setuptools 34.3.0
>
> Wha? That makes me sad, given all of the discussion on the distutils-sig
about discouraging this sort of thing.

I wonder what happens when use-dependency-links is false.

> Anyway, I am totally distracting this issue from a discussion about a
> feature into a debugging session about the speed of buildout or recipes.
>
IDK, it seems relevant. I think the main motivation for this feature is
speed.  I think it's complicated.  By providing a feature like this,
there's a danger of getting people to do things that complicate their
workflows but don't improve things.

**Comment by mauritsvanrees** [comment]
Yes, I have been using `newest = false`.
Ah, `use-dependency-links = false` helps.
And `install-from-cache = true` can also help, but you can of course only do that if you have everything local already.

I have this smaller buildout, with two parts for testing, one of them commented out. The remaining part uses `zc.recipe.egg`, which needs `setuptools`.

```
[buildout]
parts =
#    mkdir
    zc.recipe.egg
allow-picked-versions = false
show-picked-versions = true
# Only one of install-from-cache or offline can be true.
# install-from-cache = true
# offline = true
# newest = false
# use-dependency-links = false

[versions]
appdirs = 1.4.2
packaging = 16.8
# setuptools = 33.1.1
setuptools = 34.3.0
six = 1.10.0
z3c.recipe.mkdir = 0.6
zc.buildout = 2.8.0
pyparsing = 2.1.10
zc.recipe.egg = 2.0.3

[mkdir]
recipe = z3c.recipe.mkdir
paths = mypath

[zc.recipe.egg]
recipe = zc.recipe.egg
```

I ran it once to prep everything.  Then a second time:

```
maurits@MauritsBook:buildouttest $ date; time bin/buildout -v log-format="%(asctime)s %(message)s"
Fri  3 Mar 2017 21:24:15 CET
2017-03-03 21:24:17,765 Installing 'zc.buildout', 'setuptools'.
2017-03-03 21:24:17,767 We have the distribution that satisfies 'zc.buildout==2.8.0'.
2017-03-03 21:24:17,769 We have the distribution that satisfies 'setuptools==34.3.0'.
2017-03-03 21:24:17,769 Adding find link 'https://files.pythonhosted.org/packages/source/c/certifi/certifi-2016.9.26.tar.gz#md5=baa81e951a29958563689d868ef1064d' from setuptools 34.3.0
2017-03-03 21:24:19,869 Adding find link 'https://files.pythonhosted.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2' from setuptools 34.3.0
2017-03-03 21:24:21,980 Getting required 'appdirs==1.4.2'
2017-03-03 21:24:21,980 We have the distribution that satisfies 'appdirs==1.4.2'.
2017-03-03 21:24:21,981 Getting required 'six==1.10.0'
2017-03-03 21:24:21,981 We have the distribution that satisfies 'six==1.10.0'.
2017-03-03 21:24:21,982 Getting required 'packaging==16.8'
2017-03-03 21:24:21,982 We have the distribution that satisfies 'packaging==16.8'.
2017-03-03 21:24:21,984 Getting required 'pyparsing==2.1.10'
2017-03-03 21:24:21,985 We have the distribution that satisfies 'pyparsing==2.1.10'.
2017-03-03 21:24:22,119 Installing 'zc.recipe.egg'.
2017-03-03 21:24:22,120 We have the distribution that satisfies 'zc.recipe.egg==2.0.3'.
2017-03-03 21:24:22,159 Unused options for buildout: 'zope-directory'.
2017-03-03 21:24:22,159 Updating zc.recipe.egg.
2017-03-03 21:24:22,347 Installing 'zc.recipe.egg'.
2017-03-03 21:24:22,348 We have the distribution that satisfies 'zc.recipe.egg==2.0.3'.
2017-03-03 21:24:22,351 Getting required 'setuptools==34.3.0'
2017-03-03 21:24:22,352 We have the distribution that satisfies 'setuptools==34.3.0'.
2017-03-03 21:24:22,352 Adding find link 'https://files.pythonhosted.org/packages/source/c/certifi/certifi-2016.9.26.tar.gz#md5=baa81e951a29958563689d868ef1064d' from setuptools 34.3.0
2017-03-03 21:24:22,352 Adding find link 'https://files.pythonhosted.org/packages/source/w/wincertstore/wincertstore-0.2.zip#md5=ae728f2f007185648d0c7a8679b361e2' from setuptools 34.3.0
2017-03-03 21:24:22,355 Getting required 'zc.buildout==2.8.0'
2017-03-03 21:24:22,355 We have the distribution that satisfies 'zc.buildout==2.8.0'.
2017-03-03 21:24:22,357 Getting required 'appdirs==1.4.2'
2017-03-03 21:24:22,357 We have the distribution that satisfies 'appdirs==1.4.2'.
2017-03-03 21:24:22,358 Getting required 'six==1.10.0'
2017-03-03 21:24:22,358 We have the distribution that satisfies 'six==1.10.0'.
2017-03-03 21:24:22,359 Getting required 'packaging==16.8'
2017-03-03 21:24:22,359 We have the distribution that satisfies 'packaging==16.8'.
2017-03-03 21:24:22,361 Getting required 'pyparsing==2.1.10'
2017-03-03 21:24:22,362 We have the distribution that satisfies 'pyparsing==2.1.10'.

real	0m7.191s
user	0m6.680s
sys	0m0.498s
```

Then I tried the various options:

- Defaults, as above (`install-from-cache = false`, `offline = false`, `newest = true`, `use-dependency-links = true`): 7.2 seconds.
- `install-from-cache = true`: 2.7 seconds.
- `offline = true`: 6.7 seconds.
- `newest = false`: 6.8 seconds.
- `use-dependency-links = false`: 2.8 seconds.

The differences in tenths of a second are not significant.
Using multiple options does not help (or at most one tenth of a second).

Running `install instance` on my bigger buildout drops from 16 to 12 seconds when I use either `install-from-cache = true` or  `use-dependency-links = false`. The combination does not help extra.

I am going to put `use-dependency-links = false` in my `~/.buildout/default.cfg`. It might be a nice default to change in buildout 3.
